go/offline/rpc_cache.go

```diff
@@ -0,0 +1,128 @@
+package offline
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"github.com/keybase/client/go/chat/storage"
+	"github.com/keybase/client/go/encrypteddb"
+	"github.com/keybase/client/go/libkb"
+	"github.com/keybase/client/go/protocol/keybase1"
+	"golang.org/x/net/context"
+	"sync"
+)
+
+type RPCCache struct {
+	sync.Mutex
+	edb *encrypteddb.EncryptedDB
+}
+
+func newEncryptedDB(g *libkb.GlobalContext) *encrypteddb.EncryptedDB {
+	keyFn := func(ctx context.Context) ([32]byte, error) {
+		return storage.GetSecretBoxKey(ctx, g, storage.DefaultSecretUI)
+	}
+	dbFn := func(g *libkb.GlobalContext) *libkb.JSONLocalDb {
+		return g.LocalDb
+	}
+	return encrypteddb.New(g, dbFn, keyFn)
+}
+
+func NewRPCCache(g *libkb.GlobalContext) *RPCCache {
+	return &RPCCache{
+		edb: newEncryptedDB(g),
+	}
+}
+
+func hash(rpcName string, arg interface{}) ([]byte, error) {
+	h := sha256.New()
+	h.Write([]byte(rpcName))
+	h.Write([]byte{0})
+	raw, err := libkb.MsgpackEncode(arg)
+	if err != nil {
+		return nil, err
+	}
+	h.Write(raw)
+	return h.Sum(nil), nil
+}
+
+func dbKey(rpcName string, arg interface{}) (libkb.DbKey, error) {
+	raw, err := hash(rpcName, arg)
+	if err != nil {
+		return libkb.DbKey{}, err
+	}
+	return libkb.DbKey{
+		Typ: libkb.DBOfflineRPC,
+		Key: hex.EncodeToString(raw[0:16]),
+	}, nil
+
+}
+
+func (c *RPCCache) get(m libkb.MetaContext, rpcName string, encrypted bool, arg interface{}, res interface{}) (found bool, err error) {
+	defer m.CTrace(fmt.Sprintf("RPCCache#get(%s, %v, %+v)", rpcName, encrypted, arg), func() error { return err })()
+	c.Lock()
+	defer c.Unlock()
+
+	dbk, err := dbKey(rpcName, arg)
+	if err != nil {
+		return false, err
+	}
+
+	if encrypted {
+		found, err = c.edb.Get(m.Ctx(), dbk, res)
+	} else {
+		found, err = m.G().LocalDb.GetInto(res, dbk)
+	}
+	if err != nil {
+		return false, err
+	}
+	m.CDebugf("Found: %v", found)
+	return found, err
+}
+
+func (c *RPCCache) put(m libkb.MetaContext, rpcName string, encrypted bool, arg interface{}, res interface{}) (err error) {
+	defer m.CTrace(fmt.Sprintf("RPCCache#put(%s, %v, %+v)", rpcName, encrypted, arg), func() error { return err })()
+	c.Lock()
+	defer c.Unlock()
+
+	dbk, err := dbKey(rpcName, arg)
+	if err != nil {
+		return err
+	}
+
+	if encrypted {
+		err = c.edb.Put(m.Ctx(), dbk, res)
+	} else {
+		err = m.G().LocalDb.PutObj(dbk, nil, res)
```

Comment:
can you use the `GetRaw`/`PutRaw` instead so we don't json encode these? i want to cleanup the interface at some point to make it more clean how the data is stored and only use json where we need that encoding

Reply:
Thanks for looking at the PR! Is there an interface that msgpacks and then writes binary (and the opposite on the way out?)

Reply:
your change looks great. i'll cut a ticket to see if the speed/storage benefits of msgpack are worthwhile to migrate existing calls since we'll have to do some kind of migration or have all clients start with blank caches